### ISSUE TITLE
Nominatim HTTP Error 403: Forbidden for default User-Agent

### ISSUE THREAD

**Issue by xmedeko** [geopy/geopy#262]
Nominatim API seems to block the default `User-Agent: Python-urllib/3.5` now: HTTP Error 403: Forbidden.

Looking at the code the default User-Agent should be `geopy/1.11.0`?  I've tried to set own `user-agent` by: `Nominatim(user_agent='MyApp/1.2.3')` but the request has still the user agent `User-Agent: Python-urllib/3.5` and it fails with HTTP 403. So, the user-agent param does not work properly.

Edit: duplicate to #185 and probably solved by #184. Leaving this opened, since the Nominatim has started to block these requests just recently. Just to prevent other folks to duplicate the issue like me.

**Comment by xmedeko** [comment]
Note: temporary hack to overcome this issue:
```py
    def _nominatim_hack(geocoder: Nominatim):
        """HACK to geopy bugs #262 and #185"""
        requester = geocoder.urlopen

        def requester_hack(req, **kwargs):
            req = Request(url=req, headers=geocoder.headers)
            return requester(req, **kwargs)

        geocoder.urlopen = requester_hack
```

**Comment by edwin34** [comment]
![error403](https://user-images.githubusercontent.com/36815243/36642128-a479550a-1a60-11e8-9c0f-d9d35f914c5d.png)
![error403](https://user-images.githubusercontent.com/36815243/36642131-a9ef5cb4-1a60-11e8-9c59-bb7c94ee1f50.png)

**Comment by xmedeko** [comment]
@edwin34 See my "temporary hack" above.
Please, do not post screenshots of textual messages.  Furthermore, these screenshots does no add any value to the already reported issue. You can edit your post and remove them.

**Comment by xmedeko** [comment]
No, the hack does not need any file to be edited. I've just asked you to edit your post.
I've added an example usage to my post above.

**Comment by edwin34** [comment]
from geopy.geocoders import Nominatim

def nominatim_hack(geocoder: Nominatim):
        requester = geocoder.urlopen

def requester_hack(req, **kwargs):
        req = Request(url=req, headers=geocoder.headers)
        return requester(req, **kwargs)

geocoder.urlopen = requester_hack

locator = Nominatim()
locator = nominatim_hack(locator)
location = locator.reverse("52.509669, 13.376294")
print (location.address)

**This doesn't work for me too, please help me to fix this. It shows the following error:-**
 File "jio.py", line 3
    def nominatim_hack(geocoder: Nominatim):

**Comment by xmedeko** [comment]
@edwin34: invalid syntax means you have not copied the code well. The indentation level of your statements is significant in Python. You should learn it, first.

**Comment by edwin34** [comment]
![322](https://user-images.githubusercontent.com/36815243/36644217-ef863c04-1a7c-11e8-89ee-70b18f7faaa2.png)

**Comment by edwin34** [comment]
Bro, I am understand the indentation issue. But, that is not the case here. Please note that I have included only one function that is nominatim_hack, is it necessary to add requester_hack function. Can you please rectify this basic code and send me the correct code in .py format.

**Comment by eulercosta** [comment]
@edwin34
python2
```
from geopy.geocoders import Nominatim
import urllib2

def get_geolocator():
    geolocator = Nominatim()

    requester = geolocator.urlopen

    def requester_hack(req, **kwargs):
        req = urllib2.Request(url=req, headers=geolocator.headers)
        return requester(req, **kwargs)

    geolocator.urlopen = requester_hack

    return geolocator

location = get_geolocator().reverse("52.509669, 13.376294")
address = location.address
```

python3
```
from geopy.geocoders import Nominatim
from urllib.request import Request

def get_geolocator():
    geolocator = Nominatim()

    requester = geolocator.urlopen

    def requester_hack(req, **kwargs):
        req = Request(url=req, headers=geolocator.headers)
        return requester(req, **kwargs)

    geolocator.urlopen = requester_hack

    return geolocator

location = get_geolocator().reverse("52.509669, 13.376294")
address = location.address
```

**Comment by edwin34** [comment]
I am very much thankful to you Sir, (eulercosta) this actually works.

**Comment by RandyRomero** [comment]
Thank you, eulercosta! It works in geopy 1.11 perfectly. But broke down in geopy 1.12

**Comment by KostyaEsmukov** [comment]
Yeah, the reason why these hacks has broken is because the `req` argument of the `geolocator.urlopen` might be a `Request` instance rather than a string url.

This problem has been fixed in 3395aee and it's already released in `1.12.0`, so these hacks are not needed anymore. Nominatim now works out of the box again, as well as supplying a custom `user_agent`.

I'm closing this.

**Comment by Developer-Felix** [comment]
0

"GeocoderInsufficientPrivileges" error - try to use 'Photon' instead of 'Nominatim'

from geopy.geocoders import Photon
geolocator = Photon(user_agent="measurements")

I use this and it works
